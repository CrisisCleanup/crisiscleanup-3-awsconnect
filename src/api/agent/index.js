/**
 * index.js
 * Agent Module
 */

import * as Agent from './legacy';

export default Agent;
