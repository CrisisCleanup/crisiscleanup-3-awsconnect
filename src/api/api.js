/**
 * api.js
 * Base Api Module
 */

import { Dynamo } from '../utils';

export default class ApiModel {
  constructor({ loggerName = '', dbTable } = {}) {
    this.loggerName = loggerName;
    if (dbTable) {
      this.db = Dynamo.DynamoClient(dbTable);
    }
  }

  log(message, ...params) {
    console.log(`${this.loggerName} `, message, ...params);
    return this;
  }
}
