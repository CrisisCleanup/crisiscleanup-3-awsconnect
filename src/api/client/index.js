/**
 * index.js
 * Client Module
 */

import * as Client from './client';

export default Client;
