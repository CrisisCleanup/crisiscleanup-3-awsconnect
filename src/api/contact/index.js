/**
 * index.js
 * Contact Module
 */

import * as Contact from './contact';

export default Contact;
