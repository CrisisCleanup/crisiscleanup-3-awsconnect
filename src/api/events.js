/**
 * events.js
 * Events Module
 */

import axios from 'axios';

export const EVENT_OBJECTS = Object.freeze({
  AGENT: 'agent',
  INBOUND: 'inbound-call',
  OUTBOUND: 'outbound-call',
});

export const EVENT_ACTIONS = Object.freeze({
  CREATE: 'create',
  UPDATE: 'update',
});

export class Event {
  constructor({ itemId = null } = {}) {
    this.eventItem = {
      object_id: itemId,
      attr: { __source: 'awsconnect' },
    };
    this.eventDef = {
      action: null,
      object: null,
      joined: null,
    };
    this.logObjs = [];
    Object.keys(EVENT_ACTIONS).forEach((k) => {
      this[k.toLowerCase()] = () => {
        const action = EVENT_ACTIONS[k];
        this.log(`setting event action: ${action}`);
        this.eventDef.action = action;
        return this;
      };
    });
  }

  get eventKey() {
    const { object, action, joined } = this.eventDef;
    if (action === 'join' && joined) {
      // eslint-disable-next-line no-underscore-dangle
      this.eventItem.attr.__joined_id = joined.recent;
      return `${action}_${object}_to_${joined.eventDef.object}`;
    }
    return `${action}_${object}`;
  }

  get recent() {
    if (!this.logObjs.length) {
      return null;
    }
    this.log('joining recent log object:', this.logObjs);
    return this.logObjs[this.logObjs.length - 1].id;
  }

  log(message) {
    const title = `event[${this.eventItem.object_id}|${this.eventKey}]`;
    console.log(`${title} ${message}`);
  }

  object(value) {
    console.log(`setting current object: ${value}`);
    this.eventDef.object = value;
    return this;
  }

  join(event) {
    console.log(`joining event to ${event}`);
    this.eventDef.action = 'join';
    this.eventDef.joined = event;
    return this;
  }

  async save(attrs = {}) {
    this.eventItem.event_key = this.eventKey;
    this.eventItem.attr = { ...this.eventItem.attr, ...attrs };
    const resp = await axios.post('/event_logs', this.eventItem);
    this.logObjs.push(resp.data);
    return this;
  }
}
