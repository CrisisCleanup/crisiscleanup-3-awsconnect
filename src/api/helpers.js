/**
 * helpers.js
 * Api Helpers
 */

import axios from 'axios';

const LANGUAGE = {
  en_US: 2,
  es_MX: 7,
};

export const getLanguageId = async (subtag) => {
  if (subtag.includes('-')) {
    return getLanguageId(subtag.replace('-', '_'));
  }
  if (Object.keys(LANGUAGE).includes(subtag)) {
    return LANGUAGE[subtag];
  }
  return 2;
};
