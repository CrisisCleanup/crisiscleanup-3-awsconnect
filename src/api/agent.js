/* eslint-disable camelcase */
/**
 * agent.js
 * Agent Api Module
 */

import { Dynamo } from '../utils';
import { Event, EVENT_OBJECTS } from './events';

const TABLE = Dynamo.TABLES.AGENTS;

export const AGENT_STATES = Object.freeze({
  PENDING_CALL: 'PendingBusy',
  AGENT_CALLING: 'CallingCustomer',
  AGENT_PENDING: 'pending',
  ROUTABLE: 'routable',
  NOT_ROUTABLE: 'not_routable',
  BUSY: 'Busy',
  ON_CALL: 'Busy',
  OFFLINE: 'offline',
  DISCONNECTED: 'ended',
  PAUSED: 'AfterCallWork',
});

export const AGENT_STATE_TYPES = [
  AGENT_STATES.ROUTABLE,
  AGENT_STATES.NOT_ROUTABLE,
];

const ROUTABLE_STATES = [AGENT_STATES.ROUTABLE];

const NOT_ROUTABLE_STATES = [
  AGENT_STATES.ON_CALL,
  AGENT_STATES.NOT_ROUTABLE,
  AGENT_STATES.OFFLINE,
  AGENT_STATES.AGENT_PENDING,
  AGENT_STATES.PAUSED,
  AGENT_STATES.PENDING_CALL,
  AGENT_STATES.AGENT_CALLING,
];

const INROUTE_STATES = [AGENT_STATES.PENDING_CALL, AGENT_STATES.AGENT_CALLING];

export const AGENT_STATE_GROUPS = Object.freeze({
  [AGENT_STATES.ROUTABLE]: ROUTABLE_STATES,
  [AGENT_STATES.NOT_ROUTABLE]: NOT_ROUTABLE_STATES,
});

export const isRoutable = (state) =>
  ROUTABLE_STATES.includes(state.split('#')[2] || state);
export const isInRoute = (state) =>
  INROUTE_STATES.includes(state.split('#')[2] || state);

export const getStateDef = (state) => {
  if (!state || state === null || typeof state !== 'string') {
    console.log(`state ${state} is not a supported type!`);
    return getStateDef(AGENT_STATES.OFFLINE);
  }
  if (state.includes('#')) {
    console.log(`state [${state}] already contains type, returning`);
    return state.split('#');
  }
  const stateType = Object.keys(AGENT_STATE_GROUPS).filter((key) =>
    AGENT_STATE_GROUPS[key].includes(state) ? key : false,
  );
  if (stateType.length) {
    console.log(`found state type: ${stateType} for state: ${state}`);
    const isOnline = state === 'offline' ? 'offline' : 'online';
    return [isOnline, stateType[0], state];
  }
  console.log(`Unknown state type for state: ${state}`);
  return ['offline', '', state];
};

export const AGENT_ATTRS = Object.freeze({
  STATE: 'state',
  ENTERED: 'entered_timestamp',
  LAST_CONTACT: 'last_contact_id',
  CURRENT_CONTACT: 'current_contact_id',
  CONNECTION: 'connection_id',
  ACTIVE: 'active',
  STATE_TTL: 'state_ttl',
});

export const KeyMap = ({
  mapName = 'Key',
  partitionKey = TABLE.hash,
  agentId,
  attributes = {},
}) => ({
  [mapName]: {
    [partitionKey]: {
      S: agentId,
    },
    ...attributes,
  },
});

export const get = async ({ agentId, attributes }) => {
  const db = Dynamo.DynamoTable(TABLE);
  const projExp = attributes || Object.values(AGENT_ATTRS);
  const params = {
    ...KeyMap({
      agentId,
    }),
    ProjectionExpression: projExp.join(','),
  };
  if (projExp.includes('state')) {
    // state is a dynamodb reserved key word
    params.ExpressionAttributeNames = {
      '#A': 'state',
    };
    const filteredProjExp = projExp.filter((i) => i !== 'state');
    filteredProjExp.push('#A');
    params.ProjectionExpression = filteredProjExp.join(',');
  }
  console.log('fetching agent from dynamo:', params);
  const { Item } = await db.getItem(params).promise();
  if (!Item) {
    return Item;
  }
  console.log('fetched item:', Item);
  return Dynamo.normalize(Item);
};

export const setState = async ({ agentId, agentState, ...attrs }) => {
  const db = Dynamo.DynamoTable(TABLE);
  const additionalAttrs = {};
  const fetchedAgent = await get({ agentId });

  const deletedAttrs = [];
  const neededAttrs = [AGENT_ATTRS.CONNECTION];

  const [stateOnline, stateType, subState] = getStateDef(agentState);
  switch (subState) {
    case AGENT_STATES.DISCONNECTED:
    case AGENT_STATES.OFFLINE:
    case AGENT_STATES.ROUTABLE:
      if (fetchedAgent && fetchedAgent.current_contact_id) {
        console.log(
          `releasing contact id ${fetchedAgent.current_contact_id} from ${agentId}`,
        );
        deletedAttrs.push(AGENT_ATTRS.CURRENT_CONTACT);
        deletedAttrs.push(AGENT_ATTRS.STATE_TTL);
        deletedAttrs.push(AGENT_STATES.CASES);
      }
      break;
    case AGENT_STATES.AGENT_CALLING:
    case AGENT_STATES.AGENT_PENDING:
    case AGENT_STATES.PENDING_CALL:
      console.log('agent requires contact attribute! fetching!');
      neededAttrs.push(AGENT_ATTRS.CURRENT_CONTACT);
      neededAttrs.push(AGENT_ATTRS.STATE_TTL);
      neededAttrs.push(AGENT_ATTRS.CASES);
      break;
    default:
      break;
  }

  if (stateOnline) {
    additionalAttrs[AGENT_ATTRS.ACTIVE] = {
      S: 'y',
    };
  } else {
    deletedAttrs.push(AGENT_ATTRS.ACTIVE);
  }

  console.log('originally passed additional attrs:', attrs);
  const reqKeys = Object.keys(attrs);

  const finalKeys = [...new Set([...reqKeys, ...neededAttrs])].filter(
    (k) => !deletedAttrs.includes(k),
  );

  console.log('AGENT FINAL ATTRS:', finalKeys);

  const getDataType = (attr) => {
    const types = {
      number: 'N',
      string: 'S',
    };
    if (Object.keys(types).includes(typeof attr)) {
      return types[typeof attr];
    }
    return 'S';
  };

  finalKeys.forEach((key) => {
    if (attrs[key] && attrs[key] !== null) {
      additionalAttrs[key] = {
        [getDataType(attrs[key])]: attrs[key],
      };
    } else if (fetchedAgent && fetchedAgent[key]) {
      additionalAttrs[key] = {
        [getDataType(attrs[key])]: fetchedAgent[key],
      };
    }
  });
  const finalAgentState = `${stateOnline}#${stateType}#${subState}`;
  const params = {
    ...KeyMap({
      mapName: 'Item',
      agentId,
      agentState: finalAgentState,
      attributes: {
        state: {
          S: finalAgentState,
        },
        entered_timestamp: {
          S: new Date().toISOString(),
        },
        ...additionalAttrs,
      },
    }),
  };

  console.log('setting state params:', params);
  const results = await db.putItem(params).promise();
  console.log('set agent state: ', agentId, finalAgentState, results);
  const agent = Dynamo.normalize(results[0]);
  console.log('resulting set agent:', agent);

  return {
    namespace: 'phone',
    action: {
      type: 'action',
      name: 'setAgentState',
      data: {
        state: subState,
      },
    },
  };
};

export const getTargetAgent = async ({ currentContactId }) => {
  const db = Dynamo.DynamoTable(TABLE);
  const params = {
    IndexName: 'contact-index',
    ExpressionAttributeValues: {
      ':contactId': {
        S: currentContactId,
      },
    },
    KeyConditionExpression: 'current_contact_id = :contactId',
    Select: 'ALL_ATTRIBUTES',
  };
  console.log('checking for target agent...', params);
  const { Items } = await db.query(params).promise();
  if (!Items.length) {
    console.log('no current target agent!');
    return null;
  }
  const agent = Dynamo.normalize(Items[0]);
  console.log('got target agent!', agent);
  return agent;
};

export class AgentError extends Error {}

export const findNextAgent = async () => {
  const db = Dynamo.DynamoTable(TABLE);
  const params = {
    ExpressionAttributeNames: {
      '#S': 'state',
      '#O': 'active',
    },
    ExpressionAttributeValues: {
      ':t': {
        S: 'online#',
      },
      ':a': {
        S: 'y',
      },
    },
    KeyConditionExpression: '#O = :a AND begins_with (#S, :t)',
    IndexName: 'state-index',
    Select: 'ALL_ATTRIBUTES',
  };
  console.log('querying agents table with:', params);
  const { Items } = await db.query(params).promise();
  const normalizedItems = Items.map((m) => Dynamo.normalize(m));
  console.log('resulting items:', normalizedItems);
  if (!normalizedItems.length) {
    throw new AgentError('No agents are available!');
  }

  const routables = normalizedItems.filter((a) => isRoutable(a.state));
  console.log('filtered routables:');
  if (!routables.length) {
    console.log('active agents found, but none are currently routable!');
    return false;
  }

  // find the agent who's been routable the longest
  const agent = normalizedItems.reduce((pre, cur) => {
    return Date.parse(pre.entered_timestamp) > Date.parse(cur.entered_timestamp)
      ? cur
      : pre;
  });
  console.log('found longest standing routable agent:', agent);
  return agent;
};

export const createStateWSPayload = async ({ agentId, agentState }) => {
  const agent = await get({ agentId });
  return {
    namespace: 'phone',
    action: {
      type: 'action',
      name: 'setAgentState',
    },
    meta: {
      endpoint: process.env.WS_CALLBACK_URL,
      connectionId: agent.connection_id,
    },
    data: {
      state: getStateDef(agentState)[2] || getStateDef(agent.state)[2],
    },
  };
};
