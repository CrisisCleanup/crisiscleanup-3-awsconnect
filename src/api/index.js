/**
 * index.js
 * Api Module
 */

import * as Helpers from './helpers';
import * as Outbound from './outbound';

export { Helpers, Outbound };
