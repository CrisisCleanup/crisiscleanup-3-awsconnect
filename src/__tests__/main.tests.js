/**
 * main.tests.js
 * Main Tests
 */

import { Outbound } from '../api';
import Handler from '../index';

jest.mock('../api');

const MockEvent = (data = {}) => ({
  Details: {
    Parameters: {
      inboundNumber: '+10000000000',
      isDev: '0',
      action: 'CHECK_CASE',
      userLanguage: 'en_US',
      incidentId: '199',
      ...data,
    },
  },
});

describe('handler', () => {
  it('checks for existing case', async () => {
    Outbound.resolveCasesByNumber = jest.fn();
    Outbound.resolveCasesByNumber.mockResolvedValue({
      ids: [0, 1],
      pdas: [1],
      worksites: [2],
    });
    const callback = jest.fn();
    await Handler(MockEvent(), {}, callback);
    expect(Outbound.resolveCasesByNumber.mock.calls.length).toBe(1);
    expect(callback).toMatchSnapshot();
  });
});
