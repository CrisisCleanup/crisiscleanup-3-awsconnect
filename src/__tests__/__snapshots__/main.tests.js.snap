// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler checks for existing case 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      null,
      Object {
        "ids": "0,1",
        "pdas": "1",
        "worksites": "2",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`handler creates a new callback 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      null,
      Object {
        "status": "CREATED",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`handler denies a callback 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      null,
      Object {
        "status": "UNLOCKED",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`handler sets agent state 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "agentId": "xxxx",
        "agentState": "routable",
        "connection_id": null,
        "current_contact_id": null,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`handler sets agent state 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      null,
      Object {
        "promptCallType": "OUTBOUND",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
